File: packages/companion/swift/Tests/BridgethingCompanionTests/NluInferenceFake.swift
import Foundation

@testable import BridgethingCompanion

struct FakeNluInference: NluInferring {
    enum FakeError: Error { case shouldNotHaveBeenCalled }

    var logits: [String: Double] = [:]
    var inDomainLogit: Double = 8
    var slots: NluMutableSlots = .init()
    var failWithCall: Bool = false
    var logitCountOverride: Int?

    func infer(transcript: String) async throws -> NluInferenceOutput {
        if failWithCall { throw FakeError.shouldNotHaveBeenCalled }
        let count = logitCountOverride ?? NluIntentCatalog.surfaceNames.count
        var vector = Array(repeating: 0.0, count: count)
        for (name, logit) in logits {
            guard let index = NluIntentCatalog.surfaceNames.firstIndex(of: name), index < count else { continue }
            vector[index] = logit
        }
        return NluInferenceOutput(intentLogits: vector, inDomainLogit: inDomainLogit, slots: slots)
    }
}

final class PrewarmableNluInference: NluInferring, NluPrewarmable, @unchecked Sendable {
    let warmed = CallCount()

    func prewarm() async { warmed.bump() }

    func infer(transcript _: String) async throws -> NluInferenceOutput {
        NluInferenceOutput(
            intentLogits: Array(repeating: 0.0, count: NluIntentCatalog.surfaceNames.count),
            inDomainLogit: 8
        )
    }
}

final class CallCount: @unchecked Sendable {
    private let lock = NSLock()
    private var value = 0

    func bump() {
        lock.lock()
        defer { lock.unlock() }
        value += 1
    }

    var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return value
    }
}

File: packages/companion/kotlin/companion/src/test/kotlin/com/bridgething/companion/NluInferenceFake.kt
package com.bridgething.companion

import com.bridgething.schema.NluSlots
import java.util.concurrent.atomic.AtomicInteger

class FakeNluInference(
    private val logits: Map<String, Double> = emptyMap(),
    private val inDomainLogit: Double = 8.0,
    private val slots: NluSlots = NluSlots(),
    private val failWithCall: Boolean = false,
    private val logitCountOverride: Int? = null,
) : NluInferring {
    class ShouldNotHaveBeenCalled : Exception("the model ran on an utterance the caller should have claimed")

    override suspend fun infer(transcript: String): NluInferenceOutput {
        if (failWithCall) throw ShouldNotHaveBeenCalled()
        val count = logitCountOverride ?: NluIntentCatalog.surfaceNames.size
        val vector = MutableList(count) { 0.0 }
        for ((name, logit) in logits) {
            val index = NluIntentCatalog.surfaceNames.indexOf(name)
            if (index in 0 until count) vector[index] = logit
        }
        return NluInferenceOutput(intentLogits = vector, inDomainLogit = inDomainLogit, slots = slots)
    }
}

class PrewarmableNluInference : NluInferring, NluPrewarmable {
    val warmed = AtomicInteger(0)

    override suspend fun prewarm() {
        warmed.incrementAndGet()
    }

    override suspend fun infer(transcript: String): NluInferenceOutput =
        NluInferenceOutput(
            intentLogits = List(NluIntentCatalog.surfaceNames.size) { 0.0 },
            inDomainLogit = 8.0,
        )
}
